notebooks/analyze_looping_across_lending_protocols.ipynb

```diff
@@ -0,0 +1,989 @@
+    "- The visualizations will help answer several key questions:\n",
+    "  - How many users have borrowed an asset on one protocol and deposited the asset as collateral in another protocol?\n",
+    "  - How many users have completed a loop, i.e., deposited token X as collateral, borrowed token Y, deposited Y in another protocol, and borrowed X again?\n",
+    "  - What is the total dollar amount of tokens involved in these loops? How much are the deposits multiplied?\n",
+    "  - Which protocols are most subject to looping behavior? How do they compare on a per-token basis?\n",
+    "\n",
+    "# Analysis and Insights\n",
+    "The analysis will not only address the predefined questions but also explore additional metrics and hypotheses that may arise during the investigation.\n",
+    "Meaningful outputs and insights will be provided, documenting the findings and their implications for the lending protocols."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Loading the data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### From local Database"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Postgres"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
+    "import psycopg2\n",
+    "from IPython.display import display  # Only needed for Jupyter Notebook\n",
+    "\n",
+    "# Connect to the PostgreSQL database\n",
+    "conn = psycopg2.connect(\n",
+    "    host='your_host',\n",
+    "    user='your_username',\n",
+    "    password='your_password',\n",
+    "    dbname='loans_db'\n",
+    ")\n",
+    "\n",
+    "# List of protocols (table names in the PostgreSQL database)\n",
+    "protocols = [\"zklend\", \"nostra_alpha\", \"nostra_mainnet\", \"hashstack_v0\", \"hashstack_v1\"]\n",
+    "\n",
+    "for protocol in protocols:\n",
+    "    print(f\"Processing {protocol}...\")\n",
+    "    \n",
+    "    # Query the data from the PostgreSQL database\n",
+    "    query = f\"SELECT * FROM {protocol}\"\n",
+    "    df = pd.read_sql_query(query, conn)\n",
+    "    \n",
+    "    pd.set_option('display.max_columns', None)  # Display all columns\n",
+    "    pd.set_option('display.max_colwidth', None)  # Display full column width\n",
+    "    pd.set_option('display.width', None)  # Adjust display width\n",
+    "\n",
+    "    # Display the first rows\n",
+    "    display(df.head())\n",
+    "\n",
+    "# Close the connection\n",
+    "conn.close()\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### MySQL"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
```

Comment:
Move these to the imports section above.

Reply:
Done! ✅